_posts/2023-06-04-not-generating-constructors.md

```diff
@@ -0,0 +1,411 @@
+anti-pattern in Chromium and working to fix them.
+
+So it appears at first glance to be a win to change these functions to take a string view type, in
+this case `base::StringPiece`. For callers with a `std::string`, it will copy the pointer and length
+into the `base::StringPiece`. For callers with a bounded string view type, they don't need to throw
+away the bounds or heap allocate. Yet this created a pretty large binary size regression for doing
+almost nothing, and trying to eliminate heap allocations.
+
+Chromium is an interesting test bed because it is an _enormous_ codebase. So when you make a small
+change, it can easily be amplified immensely. In this case, we discovered that all those calls to a
+`base::StringPiece` constructor grow the binary by 36KB. For unknown reasons, the
+`base::StringPiece` constructor gets inlined, while the `std::string_view` constructor does not, so
+you get less binary size growth with that type (and we'll look at marking `base::StringPiece` with a
+`noinline` attribute).
+
+But in our experience, we've seen that binary size correlates quite well with performance as well.
+So *this also implies a performance degredation by moving from pointers to a bounded view type*. And
+herein lies the problem. Even if a codebase successfully flip libc++ hardening on for their
+`std::span` types because the majority of bounds checks are elided, following through on converting
+their pointers to `std::span` will introduce a different kind of performance (and binary size)
+regression.
+
+Here's a code example ([in compiler explorer](https://godbolt.org/z/14Kq9GG3r)):
+
+```cpp
+#include <stddef.h>
+#include <vector>
+#include <span>
+#include <fmt/core.h>
+
+void take_vector(const std::vector<int>& v) {
+    fmt::println("{}", v[0]);
+}
+
+void take_pointer(const int* p, size_t s) {
+    if (0 >= s) abort();
+    fmt::println("{}", p[0]);
+}
+
+void take_span(std::span<const int> s) {
+    fmt::println("{}", s[0]);
+}
+
+int main() {
+    std::vector<int> a = {1, 2, 3, 4, 5, 6, 7};
+    take_vector(a);
+    take_pointer(a.data(), a.size());
+    take_span(a);
+}
+```
+
+We will look at the resulting code gen unoptimized, to keep things simple. But the above commit
+demonstrated the same principle under optimizations.
+
+Here's the call to `take_vector()`:
+```asm
+lea     rax, [rbp-80]
+mov     rdi, rax
+call    take_vector(std::vector<int, std::allocator<int> > const&)
+```
+This is three instructions. Grab the pointer, store it, and call the function. Nice.
+
+Here's the call to `take_pointer()`:
+```asm
+lea     rax, [rbp-80]
+mov     rdi, rax
+call    std::vector<int, std::allocator<int> >::size() const
+mov     rbx, rax
+lea     rax, [rbp-80]
+mov     rdi, rax
+call    std::vector<int, std::allocator<int> >::data()
+mov     rsi, rbx
+mov     rdi, rax
+call    take_pointer(int const*, unsigned long)
+```
+That's a lot more instructions, as there's 2 function calls first. Then store the two values and

```

Comment:
"Then the program stores the 2 values and makes the call"

(be consistent about using either numerals or number words)

Reply:
I always write numerals but want to use words lol